compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -44,23 +44,15 @@ pub fn check_legal_trait_for_method_call(
     trait_id: DefId,
 ) {
     if tcx.lang_items().drop_trait() == Some(trait_id) {
-        let mut err = struct_span_err!(tcx.sess, span, E0040, "explicit use of destructor method");
-        err.span_label(span, "explicit destructor calls not allowed");
-
-        let (sp, suggestion) = receiver
-            .and_then(|s| tcx.sess.source_map().span_to_snippet(s).ok())
-            .filter(|snippet| !snippet.is_empty())
-            .map(|snippet| (expr_span, format!("drop({snippet})")))
-            .unwrap_or_else(|| (span, "drop".to_string()));
-
-        err.span_suggestion(
-            sp,
-            "consider using `drop` function",
-            suggestion,
-            Applicability::MaybeIncorrect,
-        );
-
-        err.emit();
+        let sugg = if receiver.is_some_and(|s| !s.is_empty()) {
+            errors::ExplicitDestructorCallSugg::Snippet {
+                lo: expr_span.shrink_to_lo(),
+                hi: expr_span.shrink_to_hi(),
```

Comment:
This doesn't change `x.drop()` into `drop(x)` -- it changes it to `drop(x.drop())`. Perhaps something like this works (using an `if let` instead of `is_some_and` above):

```suggestion
                hi: receiver.shrink_to_hi().to(expr_span.shrink_to_hi()),
```
?

So we can replace this:

```
x.drop()
 ^^^^^^^
```